pyramid_swagger/load_schema.py

```diff
@@ -44,100 +46,85 @@ def build_param_schema(schema, param_type):
-def extract_body_schema(schema, models_schema):
-    """Turn a swagger endpoint schema into an equivalent one to validate our
-    request.
+def filter_params_by_type(schema, param_type):
+    return [s for s in schema['parameters'] if s['paramType'] == param_type]
 
-    As an example, this would take this swagger schema:
-        {
-            "paramType": "body",
-            "name": "body",
-            "description": "json list: [ll1,ll2]",
-            "type": "array",
-            "items": {
-                "$ref": "GeoPoint"
-            },
-            "required": true
-        }
-    To this jsonschema:
-        {
-            "type": "array",
-            "items": {
-                "$ref": "GeoPoint"
-            },
-        }
-    Which we can then validate against a JSON object we construct from the
-    pyramid request.
-    """
-    matching_body_schemas = [
-        s
-        for s in schema['parameters']
-        if s['paramType'] == 'body'
-    ]
-    if matching_body_schemas:
-        schema = matching_body_schemas[0]
-        type_ref = extract_validatable_type(schema['type'], models_schema)
-        # Unpleasant, but we are forced to replace 'type' defns with proper
-        # jsonschema refs.
-        if '$ref' in type_ref:
-            del schema['type']
-            schema.update(type_ref)
-        return strip_swagger_markup(schema)
-    else:
-        return None
 
+def extract_body_schema(schema):
+    """Return the body parameter schema from an operation schema."""
+    matching_body_schemas = filter_params_by_type(schema, 'body')
+    # There can be only one body param
+    return matching_body_schemas[0] if matching_body_schemas else None
 
-# TODO: do this with jsonschema directly
-def strip_swagger_markup(schema):
-    """Turn a swagger URL parameter schema into a raw jsonschema.
 
-    Involves just removing various swagger-specific markup tags.
+def ignore(_validator, *args):
+    """A validator which performs no validation. Used to `ignore` some schema
+    fields during validation.
     """
-    swagger_specific_keys = (
-        'paramType',
-        'name',
-    )
-    return dict(
-        (k, v)
-        for k, v in schema.items()
-        if k not in swagger_specific_keys
-    )
+    return
+
+
+def build_swagger_type_validator(models):
+    def swagger_type_validator(validator, ref, instance, schema):
+        func = _validators.ref if ref in models else _validators.type_draft4
+        return func(validator, ref, instance, schema)
+
+    return swagger_type_validator
 
 
-def get_model_resolver(schema):
+def required_validator(validator, req, instance, schema):
+    """Swagger 1.2 expects `required` to be a bool in the Parameter object, but
+    a list of properties in a Model object.
     """
-    Get a RefResolver. RefResolver's will resolve "$ref:
-    ObjectType" entries in the schema, which are used to describe more complex
-    objects.
+    if schema.get('paramType') and req is True:
```

Comment:
If `schema.get('paramType')` is `True` but `req` is `False`, in that case it should not call `required_draft4` with `req` with a boolean value, as the method expects a list.